enhancements/console/customize-perspectives.md

```diff
@@ -0,0 +1,294 @@
+is responsible for managing the cluster resources, users, and projects.
+
+This document provides details about how cluster administrator 
+can hide user perspectives. 
+
+![Perspectives switch dropdown](perspectives-switch.png)
+
+## Motivation
+
+Cluster administrator wants to hide the "Administrator" perspective 
+from non-privileged and to hide the "Developer" 
+perspective for all users.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want to hide "Administrator" perspective 
+from all non-privileged users so that non-privileged users
+cannot access "Administrator" perspective for managing the cluster resources, 
+users, and projects. Non-privileged users are users that cannot get
+all `namespaces` (based on RBAC check)
+
+#### Story 2
+
+As a cluster admin, I want to hide "Developer" perspective
+for all users.
+
+#### Story 3
+
+As a cluster admin, I want to hide any perspective contributed by a plugin.
+
+Additional perspectives can be dynamically added with the 
+[Perspective extension](https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/extensions/perspectives.ts).
+
+### Goals
+
+- Allow cluster admins to hide the user perspectives.
+- Provide a way that cluster admins can do this based on resource permissions (RBAC).
+
+### Non-Goals
+
+- Allow cluster admins to customize the navigation menu along with perspectives for users or groups.
+
+## Proposal
+
+We can extend the existing `operator.openshift.io/v1` / `Console` CRD. 
+We can add 
+`spec.customization.perspectives` which can have list of customization details.
+
+### Workflow Description
+
+To disable perspective for users with limited permissions we need a list of options with a perspective id and the required permissions.
+
+```text
+├── Spec
+│   ├── customization
+|       ├── perspectives 
+|           ├── id
+|           ├── requiresAccessReview
+|               ├── kind
+|               ├── verb
+└── ...
+```
+
+This might be extendable in the future if we want replace (hide one, show another) a perspective based on a resource permissions:
+
+```text
+├── Spec
+│   ├── customization
+|       ├── perspectives 
+|           ├── id
+|           ├── requiresAccessReview
+|           |   ├── kind
+|           |   ├── verb
+|           ├── missingAccessReview
+|               ├── kind
+|               ├── verb
+└── ...
+```
+
+Each perspective conforms to the following schema:
+
+- `id: string` , defines the id of perspective to disable. Incorrect or unknown ids will be ignored.
+
+// +optional
+- `requiresAccessReview: authV1.ResourceAttributes[]` , defines a list of permission checks. The perspective will only be shown when all checks are successful.
+
+// +optional
+- `missingAccessReview: authV1.ResourceAttributes[]` , defines a list of permission checks. The perspective will only be shown when at least one check fails.
+
+// +optional
+- `state: Enabled/Disabled` , defines the perspective is enabled or disabled.
```

Comment:
How does this interact with the missing and required permissions?

Reply:
for `missingAccessReview` see above.

for `state: Disabled`: This will always disable the perspective, independent from `requiresAccessReview` and `missingAccessReview`. An admission controller could check this, but we don't have one yet.

Reply:
Right, in which case, we probably want to make this a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions), so you'd have something like
```
- id: <blah>
  visibility:
    state: Enabled # Always show this perspective
- id: <blah>
  visibility:
    state: Disabled # Never show this perspective
- id: <blah>
  visibility:
    state: AccessReview # Show the perspective based on the access review defined
    accessReview: # This section defines how the access review should work
      required: # Users then define the required permissions
         - <perms>
     missing: # And the missing permissions, and this is and'ed with the required(?)
         - <perms>
```
I'm also wondering if we need to have some sort of `AllOf`/`AnyOf` concept within the permission but we can discuss that separately

Reply:
Added the above structure by adding `visibility`. Thanks.